sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2274,33 +2274,41 @@ case class Concat(children: Seq[Expression]) extends ComplexTypeMergingExpressio
 
   override def foldable: Boolean = children.forall(_.foldable)
 
-  override def eval(input: InternalRow): Any = dataType match {
+  override def eval(input: InternalRow): Any = evalFunction(input)
+
+  @transient private lazy val evalFunction: InternalRow => Any = dataType match {
```

Comment:
We don't `@transient` here? For example, the similar logic has no transient: https://github.com/apache/spark/blob/47d6e80a2e64823fabb596503fb6a6cc6f51f713/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala#L1252

Reply:
Also, how about `evalFunction` -> `concatFunction` or `doConcat`?

Reply:
i think ti's good to have a transient to reduce closure size